### DISCUSSION TITLE
Is SyncProducer faster for generating a single message per request compared to AsyncProducer

### DISCUSSION THREAD

**Discussion by abhishekgahlot2** [IBM/sarama#2806] [Q&A]
I am trying to optimize event processing 

Here is my code

```
func setupProducer() (sarama.SyncProducer, error) {
	config := sarama.NewConfig()
	config.Net.SASL.Enable = true
	config.Net.SASL.Mechanism = sarama.SASLTypeOAuth
	config.Net.SASL.TokenProvider = &MSKAccessTokenProvider{}

	tlsConfig := tls.Config{}
	config.Net.TLS.Enable = true
	config.Net.TLS.Config = &tlsConfig
	config.Producer.Return.Successes = true
	return sarama.NewSyncProducer(kafkaBrokers, config)
}


func produceMessages(producer sarama.SyncProducer, signals chan os.Signal, KafkaTopic string, Value string) {
	log.Println(KafkaTopic, Value)
	message := &sarama.ProducerMessage{Topic: KafkaTopic, Value: sarama.StringEncoder(Value)}
	producer.SendMessage(message)
}
```

I want to produce message of maximum payload possible and end request. I am considering using AsyncProducer here but not sure if its going to be of bigger help.

**Comment by dnwe** [comment]
Great question! Yes, AsyncProducer will typically give you better throughput than SyncProducer because it is able to run freely accepting messages to send and batching them into requests. SyncProducer is really just a convenient [wrapper](https://github.com/IBM/sarama/blob/fd84c2b0f0185100dbaec28ca4074289b35cc1b1/sync_producer.go#L60-L76) around an AsyncProducer to make the API easier to consume in a synchronous app. 

However, when you say “generate single message per request“ do you mean that you want to produce a single Message to any given partition in Kafka and to not send the next Message until you’ve received an Acknowledgement for that previous message?
